File: reglaFalsa.py
# 
# reglaFalsa.py
#
# @Author Jose Alfredo Núñez Aguirre - 000000181930
#
# Programa que encuentra las Raices utilizando el método de la Regla Falsa 
# de la funcion f(x) = x^4 - 8x^3 -  35x^2 + 450x - 1001, 


print("\033[4;35m")
print('* * Programa que encuentra las Raices utilizando el método de Biseccion')
print('de la funcion función f(x) = sen(x) - x^2, * * ')
print("\033[0;32m")

# Método que recibe una variable x para calcular la función  f(x) = x^4 - 8x^3 -  35x^2 + 450x - 1001,
# y devuelve el resultado como flotante
def f (x):

    return x**4 - 8 * (x**3) - 35*(x**2) + 450*(x) - 1001



# Método que calcula la raíz de la funcion  através de REGLA FALSA 
# xi: Representa el valor inicial 
# xf: Representa el valor final
# errMAx: representa el maximo error 
def reglaFalsa(xi,xf,errMax): 

    # Variable que guarda el valor de cada iteracion empezando de 0
    iterador = 0

    # vector donde se almacena el valro de xr actual de acuerdo al iterador
    actual = []

    # variable error que guarda el error con punto flotante
    error = 0

    # variable que guarda el el error multiplicado * 100 y lo convierte en un valor absoluto
    errorAbs =0

    while True:
       

        # se realiza el calculo de xr usando la formula xr = ( (xf - xi)  / f(xf) - f(xi)) * f(xi)
        xr = xi -  ( xf - xi)  /  ( f(xf) -  (f(xi))  ) * f(xi)

        # se guarda el valor de xr en el arreglo actual para determinar el xr anterior de acuerdo al indice menos 1    
        actual.append(xr)
        
        # se calcula la multiplicacion de la funciones:  f(xi) * f(xr)  y se guarda en la variable xixr
        xixr =f(xi) * f(xr)

        
        # Si el iterador es mayor a cero se realiza la operación que nos da el valor de el Error Absoluto
        if iterador > 0:
            
            # Se calcula el error con xrNueva - xrAnterior / xrNueva 
            # Se extrae el valor anterior del arreglo con nombre actual, deacuerdo  iterador - 1 para encontrar la posicion del valor.
            error = ( ( xr - actual[iterador -1] ) / xr ) 

            # Se convierta % y a valor absoluto
            errorAbs = abs(error) * 100

            # Formato de resultado en pantalla
            print(f'{iterador + 1}|\t{xi:>8.4f}|{xf:>8.4f}|{xr:>8.4f}|{f(xi):>8.4f}|{f(xf):8.4f}|{f(xr):8.4f}|{xixr:>12.4f}|{errorAbs:>7.4f} % | {error:>8.4f}|')
             
            # Si el error absoluto es menor a el error Maximo el programa cierra.
            if errorAbs < errMax:
               break
        
        # si la condicion no se cumple imprime en pantalla la cabezera de la tabla y la primera resolucioón donde no se calcula el error.
        else: 
    
            print('\033[1;31m')
            print(f'i|\t{"Xi":>8}|{"Xf":>8}|{"Xr":>8}|{"F(xi)":>8}|{"F(xf)":>8}|{"F(xr)":>8}| {"F(Xi)*F(xr)":>4}|{"Eamax %":>10}| {"EaMax":>8}| \033[0;37m ')
    
            print("----------------------------------------------------------------------------------------------")

            print(f'{iterador + 1}|\t{xi:>8.4f}|{xf:>8.4f}|{xr:>8.4f}|{f(xi):>8.4f}|{f(xf):8.4f}|{f(xr):8.4f}|{xixr:>12.4f}|{"N/A":>10}| {"N/A":>8}|')

        
        # si F(xi) * F(xr) es mayor a eso le asignamos el valor de xr a xi
        if xixr > 0:         
            xi = xr

        # si F(xi) * F(xr) es menor a eso le asignamos el valor de xr a xf    
        if xixr < 0:
            xf = xr

        # aumenta el valor del iterador en 1.    
        iterador+=1



inicial = float(input("Escribe el valor inicial: "))
print()
final = float(input("Escribe el valor final: "))
print()
errorMax = float(input("Escribe el valor de Error Maximo: ")) 
print("\033[0;37m")




reglaFalsa(inicial,final,errorMax)

File: reglaFalsa/src/reglafalsa/ReglaFalsa.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package reglafalsa;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.text.ParseException;

/**
 * @Class ReglaFalsa 
 * Programa que determina una de las raíces reales de la
 * función: f(x) = (1 – 0.6x)/x, empleando el Método de la Regla Falsa. El
 * programa pide los valores inicial y final del intervalo de búsqueda y el
 * error aproximado máximo. El programa calcula y despliega los valores de la
 * raíz, el valor de la función para esa raíz y el número de iteraciones
 * requerida para encontrar la raíz.
 * @author José Alfredo Nuñez Aguirre
 */
public class ReglaFalsa {

    public static void main(String args[]) throws ParseException {

        System.err.println(" * Este Programa determina una de las raíces reales de la función: f(x) = (1 – 0.6x)/x,\n"
                + " * empleando el Método de la Regla Falsa. El programa pide los valores inicial y final del\n"
                + " * intervalo de búsqueda y el error aproximado máximo. El programa calcula y despliega\n"
                + " * los valores de la raíz, el valor de la función para esa raíz y el número de iteraciones requerida\n"
                + " * para encontrar la raíz.\n*********************************************\n");

        try {

            // objeto para obtener la entrada en consola
            BufferedReader entrada = new BufferedReader(new InputStreamReader(System.in));

            System.out.print("Ingresa el valor x inicial: ");
            double xi = Double.parseDouble(entrada.readLine());

            System.out.print("Ingresa el valor x final: ");
            double xd = Double.parseDouble(entrada.readLine());

            System.out.print("Ingresa el valor del Error: ");
            double error = Double.parseDouble(entrada.readLine());

            // se llama al metodo reglaFalsa para realizar la busque la raiz
            reglaFalsa(xi, xd, error);

        } catch (IOException e) {
            System.out.println(e.getMessage());
        } catch (NumberFormatException e) {
            System.err.println("Solo se aceptan números de tipo Double");
        }

    }

    /**
     * Metodo f que calcula la ecuacion (1 – 0.6x)/x
     *
     * @param x representa el valor a sustituir en X
     * @return devuelve el resultado de tipo double obtenido en la ecuacion.
     */
    public static double f(double x) {

        return (1 - (0.6 * x)) / x;

    }

    /**
     * Metodo reglaFalsa, busca las raices de la ecuacion f(x)=(1 – 0.6x)/x e
     * imprime una tabla.
     *
     * @param xi Representa el valor de X izquierda o X inicial.
     * @param xd Representa el valor de X derecha o X final
     * @param error Representa el valor del error en decimales.
     */
    public static void reglaFalsa(double xi, double xd, double error) {

        // variable fxi donde se guarda el valor de el resultado de f(xi);
        double fxi;
        
        // variable fxd donde se guarda el valor de el resultado de f(xd);
        double fxd;
        
        // variable xm /= ((xi * fxd) - (xd * fxi)) / (fxd - fxi);
        double xm;
        
        // variable fxm donde se guarda el valor de el resultado de f(xm) se inicializa en uno como auxiliar 
        //para que entre a while y dentro de while se calcula el primer resultado de manera correcta.
        double fxm = 1;
        
        // contador auxiliar para el conteo de las iteraciones realizadas.
        int contador = 0;

        System.out.printf("%-11s | %-9s | %-9s | %-9s | %-9s | %-9s | %-9s\n",
                "interacion", "xini", "xfin", "xm", "f(xini)", "f(xfin)", "f(xm)");
        System.out.println("-------------------------------------------------------------------------------------");

        // mientras el valor absoluto de xm sea mayor que el error  continuan las operaciones.
        while (Math.abs(fxm) > error) {

            // se calcula f(xi) con la funcion f(x)
            fxi = f(xi);
            
            // se calcula f(xd) con la funcion f(x)
            fxd = f(xd);
            
            // se calcula xm con la ecuacion (xi * fxd) - (xd * fxi)) / (fxd - fxi)
            xm = ((xi * fxd) - (xd * fxi)) / (fxd - fxi);

            // se calcula el f(xm) con la funcion f(x)
            fxm = f(xm);

            // formato de la tabla usando 4 decimales.
            System.out.format("%-11d | %9.4f | %9.4f | %9.4f | %9.4f | %9.4f | %9.4f |  \n",
                    contador, xi, xd, xm, fxi, fxd, fxm);
            System.out.println("-------------------------------------------------------------------------------------");

            // Si f(xm) es menor a 0 entonces el valor de xm pasa a ser el nuevo valor de xi
            if (fxm < 0) {

                if (fxi < 0) {
                    xi = xm;
                } else {
                    xd = xm;
                }

            } else {
                // en caso contrario xm pasa aser el nuevo valor de xd
                if (fxi > 0) {
                    xi = xm;
                } else {
                    xd = xm;
                }

            }
            // se aumenta el número de iteraciones
            contador++;

        }

    }

}
